Move compilation into /lib

roc opened the pull request:
### Adds in test fixtures
- English and Arabic `src` folders (culled from EVW)
- any outputted files are ignored in `.gitignore`
### lib/compile.js
- Adds callback to the end of compilation
  - so we can play around with post-walk actions
  - also good for tests passing `done()`
- Removes dependency on `underscore-deep-extend`
  - Object.assign is fine here
- Separates `directory` function out to reduce nesting in callbacks
- Adds tests
  - dumb tests, there to bootstrap in an end to end fixture related test
  - can get more granular as the compiler code becomes more granular
### Removes use of `rimraf`
- we don't need it, as `mkdirp` is overwriting our folders
- `dirs.delete` left in, ready to be deprecated/removed unless there's some terrible need for it that I've missed
### Remove `underscore` dependency
- iterating and combining objects using native `Object` methods

roc committed:
Add in lib/dirs, for directory operations

- only deletion for now

roc committed:
Just pop the array in getLanguage

roc committed:
Make a directory function, call it

roc committed:
No longer deleting, no longer underscoreDeepExtending

- `Object.assign` works just fine for composing objects
- no need to delete folders, as we're always overwriting

roc committed:
Add in test fixtures

- English and Arabic (working on evw at the moment)
- any outputted files are ignored in .gitignore

roc committed:
Move compilation to own file (lib/compile.js)

A big heave, this.

- Adds callback to the end of compilation
  - so we can play around with post-walk actions
  - also good for tests passing `done()`
- Removes dependency on `underscore-deep-extend`
  - Object.assign is fine here
- Separates `directory` function out to reduce nesting in callbacks
- Adds tests
  - dumb tests, there to bootstrap in an end to end fixture related test
  - can get more granular as the compiler code becomes more granular
- Removes use of rimraf entirely

roc committed:
Remove underscore dependency entirely

- iterating and combining objects using native `Object` methods

joefitter commented:
nice, LGTM - @JoeChapman ?

roc committed:
test/compile => test/lib/compile

- Also adds a timeout wrapper around pulling out all of the compiled fixture results
- This is a bit boneheaded, but I feel it's a useful integration style testing start before more granular testing is added

roc commented:
@joefitter @JoeChapman I'm doing a slightly ridiculous thing in the tests (but it's only in the tests) here: https://github.com/UKHomeOffice/hof-transpiler/pull/24/files#diff-bf854085839aa797e5cd956dbc1cc6d7R35

Feel like this is still a good patch before someone comes along and does a clarifying refactor at some point

joefitter commented:
I did see that - so you are checking the contents of the compiled file? Seems fine for tests as we want to isolate functionality as much as poss.

roc commented:
@joefitter Yeah checking the folders exist, then checking the compiled `default.json` files look correct. Nothing fancy. The compile callback is fired immediately after the `walker.walk` sequence has finished, but the read/stat-ing seems to take a little longer, hence the slightly resigned timeout wrapper.

joefitter commented on lib/compile.js:
```diff
@@ -0,0 +1,105 @@
+      });
+    });
+  });
+  if (cb) {
```
@roc could it be that we are calling cb outside of the walker.on end event?

roc replied:
@joefitter I was just writing a very long explanation of why that's not the case, but I think I must've fudged the `cb` placement originally, and actually moving it to the bottom of the `'end'` statement works a treat. Who would've thought?!

joefitter replied:
always something simple ;)

roc replied:
here we go https://github.com/UKHomeOffice/hof-transpiler/pull/24/commits/f624d73c48e0c6f0867b3b721883279ed5b79395

roc committed:
Put compile callback in the callback of `walker.on('end')`

- Was errantly placed at the bottom of the compile function before, so was firing before the walking had finished
- now it's wrapped in the `walker.walk` sequence, firing in the callback of `walker.end`
- allows us to be more confident about writes having finished

joefitter commented:
boom - lgtm @JoeChapman ?

JoeChapman commented:
looking

JoeChapman commented on .gitignore:
```diff
@@ -1,3 +1,5 @@
 .idea/*
+test/fixtures/ar
+test/fixtures/en
```
maybe just `test/fixtures/**`?

joefitter replied:
test/fixtures/src is needed - merging

JoeChapman commented:
LGTM

roc commented:
🎉 psyche🎉
